crates/revm/src/db/alloydb.rs

```diff
@@ -0,0 +1,181 @@
+            let balance = self.provider.get_balance(address, self.block_number);
+            let code = self
+                .provider
+                .get_code_at(address, self.block_number.unwrap_or_default());
+            tokio::join!(nonce, balance, code)
+        };
+
+        let (nonce, balance, code) = Self::block_on(f);
+
+        let balance = balance?;
+        let code = Bytecode::new_raw(code?.0.into());
+        let code_hash = code.hash_slow();
+        let nonce = nonce?;
+
+        Ok(Some(AccountInfo::new(
+            balance,
+            nonce.to::<u64>(),
+            code_hash,
+            code,
+        )))
+    }
+
+    fn block_hash_ref(&self, number: U256) -> Result<B256, Self::Error> {
+        // Saturate usize
+        if number > U256::from(u64::MAX) {
+            return Ok(KECCAK_EMPTY);
+        }
+
+        let block = Self::block_on(
+            self.provider
+                // SAFETY: We know number <= u64::MAX, so we can safely convert it to u64
+                .get_block_by_number(number.to::<u64>().into(), false),
+        )?;
+        // SAFETY: If the number is given, the block is supposed to be finalized, so unwrapping is safe.
+        Ok(B256::new(*block.unwrap().header.hash.unwrap()))
+    }
+
+    fn code_by_hash_ref(&self, _code_hash: B256) -> Result<Bytecode, Self::Error> {
+        panic!("This should not be called, as the code is already loaded");
+        // This is not needed, as the code is already loaded with basic_ref
+    }
+
+    fn storage_ref(&self, address: Address, index: U256) -> Result<U256, Self::Error> {
+        let slot_val = Self::block_on(self.provider.get_storage_at(
+            address,
+            index,
+            self.block_number,
+        ))?;
+        Ok(slot_val)
+    }
+}
+
+impl<P: Provider> Database for AlloyDB<P> {
+    type Error = TransportError;
+
+    #[inline]
+    fn basic(&mut self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
+        <Self as DatabaseRef>::basic_ref(self, address)
+    }
+
+    #[inline]
+    fn code_by_hash(&mut self, code_hash: B256) -> Result<Bytecode, Self::Error> {
+        <Self as DatabaseRef>::code_by_hash_ref(self, code_hash)
+    }
+
+    #[inline]
+    fn storage(&mut self, address: Address, index: U256) -> Result<U256, Self::Error> {
+        <Self as DatabaseRef>::storage_ref(self, address, index)
+    }
+
+    #[inline]
+    fn block_hash(&mut self, number: U256) -> Result<B256, Self::Error> {
+        <Self as DatabaseRef>::block_hash_ref(self, number)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use alloy_provider::ProviderBuilder;
+
+    use super::*;
+
+    #[test]
+    fn can_get_basic() {
+        let client = ProviderBuilder::new()
+            .on_reqwest_http(
+                "https://mainnet.infura.io/v3/c60b0bb42f8a4c6481ecd229eddaca27"
+                    .parse()
+                    .unwrap(),
+            )
+            .unwrap()
+            .boxed();
```

Comment:
boxed not needed i think

Reply:
Ah yeah, this is an interesting byproduct of the `<T, N>` bounds on `Provider<T, N>`:

In _this_ case, this was needed because `AlloyDB<P: Provider>` was defaulting to `Provider<T = BoxTransport, N = Ethereum>`, therefore it was expecting a boxed transport. 

To make this fully generic we need to make `AlloyDb` be `pub struct AlloyDB<T: Transport + Clone, N: Network, P: Provider<T, N>>` So we can make `T: Transport + Clone, N: Network` and therefore accept any transport, and any network. The annoying issue about this is that rustc complains about `T` and `N` being "unused" because they're not in the body, so we have to mark them as used with phantom data markers. See https://github.com/rust-lang/rust/issues/23246